Feature: make the number of columns that are formatted as a table in Teams alerts a cli flag so that users have more control over the formatting in Teams

FrankTub opened the pull request:
I've played around quite a lot with this teams integrations and sadly it is indeed not possible to detect for the new workflows alert webhooks to determine if an alert was sent successfully.

By default the formatting of the tables will behave as it used to do with one exception. The full width is used for the message, independent of the number of columns that the Test Results Sample will contain. So in the old situation:

<img width="940" height="738" alt="image" src="https://github.com/user-attachments/assets/76372f0f-2cc5-4319-b969-3491dae99584" />

In the new situation it will always look like:

<img width="966" height="425" alt="image" src="https://github.com/user-attachments/assets/0514f45d-e885-4007-b525-722252dfed89" />

When running something like it will simply behave like it used to:

```bash
edr monitor --teams-webhook "$TEAMS_WEBHOOK_URL" --report-url "$DATA_QUALITY_REPORT_URL" --profile-target default
```

The new situation can be triggered with for example:

```bash
edr monitor --teams-webhook "$TEAMS_WEBHOOK_URL" --report-url "$DATA_QUALITY_REPORT_URL" --profile-target default -mc 10
```

Example output:

<img width="922" height="684" alt="image" src="https://github.com/user-attachments/assets/fcee5e34-0269-4d08-937a-66d31b342aaa" />
<!-- pylon-ticket-id: 57a908c7-04c9-4c91-9772-1479b957936d -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * New config/CLI option to set the maximum columns shown in alert samples; display switches between table and JSON accordingly.
* **Improvements**
  * Adaptive Cards render full-width in Microsoft Teams, wrap long table cell text, and avoid showing "None" for empty cells.
* **Bug Fixes**
  * More reliable Teams webhook handling with clearer success criteria, improved error messages, and rate limiting.
* **Refactor**
  * Reuse a single, configured message builder for consistent alert formatting.
* **Tests**
  * Updated Adaptive Card fixtures for Teams width and text wrapping.
* **Chores**
  * Added ignore rule for internal dbt packages.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

michael-myaskovsky commented on elementary/messages/messaging_integrations/teams_webhook.py:
```diff
@@ -56,23 +61,31 @@ def __init__(self, url: str) -> None:
-        except requests.RequestException as e:
+        response = send_adaptive_card(self.url, card)
+        # The status code is not reliable for the older version and 202 for the newer version
+        # The response returns 1 in the body for the old version for success and otherwise sometext
+        # The new version is not reliable to do a check and in the response there is nothing that can be used
+        # to determine if the message was sent successfully.
+        if (
+            response.status_code not in (HTTPStatus.OK, HTTPStatus.ACCEPTED)
+            or len(response.text) > 1
+        ):
             raise MessagingIntegrationError(
-                "Failed to send message to Teams webhook"
-            ) from e
+                f"Could not post message to Teams via webhook - {self.url}. Status code: {response.status_code}, Error: {response.text}"
+            )
+        return MessageSendResult(
+            message_context=EmptyMessageContext(),
+            timestamp=datetime.utcnow(),
+            message_format="adaptive_cards",
+        )
```
@FrankTub The second comment here seems relevant, how would the new code handle an exception?

michael-myaskovsky commented on elementary/monitor/alerts/alert_messages/builder.py:
```diff
@@ -343,7 +344,8 @@ def _get_result_blocks(
             elif result_sample:
                 if (
                     isinstance(result_sample, list)
-                    and len(result_sample[0].keys()) <= 4
+                    and len(result_sample[0].keys())
+                    <= self.config.maximum_columns_in_alert_samples
                 ):
```
@FrankTub

michael-myaskovsky commented on elementary/monitor/data_monitoring/alerts/data_monitoring_alerts.py:
```diff
@@ -309,7 +309,7 @@ def _send_alert(
         if isinstance(self.alerts_integration, BaseIntegration):
             return self.alerts_integration.send_alert(alert)
-        alert_message_builder = AlertMessageBuilder()
+        alert_message_builder = AlertMessageBuilder(self.config)
```
@FrankTub check if the type matches

michael-myaskovsky reviewed (commented):
Hi @FrankTub! Thank you for your contribution, I left some comments under coderabbit's comments, if you could take a look and explain/fix I would be glad to approve your contribution.
Thanks again!

FrankTub commented:
@michael-myaskovsky I think I addressed all your comments. I will highlight one piece of code which I just don't get how this ever worked in this setup.

elazarlachkar commented on elementary/messages/formats/adaptive_cards.py:
```diff
@@ -131,7 +131,7 @@ def format_fact_list_block(block: FactListBlock) -> Dict[str, Any]:
 
 
-def format_table_block(block: TableBlock) -> Dict[str, Any]:
+def format_table_block(block: TableBlock) -> dict:
```
Why this change?

elazarlachkar commented on elementary/messages/messaging_integrations/teams_webhook.py:
```diff
@@ -42,8 +45,10 @@ def send_adaptive_card(webhook_url: str, card: dict) -> requests.Response:
-        logger.debug("Got 202 response from Teams webhook, assuming success")
+    if response.status_code == HTTPStatus.ACCEPTED:
+        logger.debug(
+            "Got %s response from Teams webhook, assuming success", HTTPStatus.ACCEPTED
```
Use `f` formatting here

elazarlachkar commented on elementary/messages/messaging_integrations/teams_webhook.py:
```diff
@@ -56,22 +61,35 @@ def __init__(self, url: str) -> None:
         except requests.RequestException as e:
             raise MessagingIntegrationError(
-                "Failed to send message to Teams webhook"
+                f"Network error while posting message to Teams webhook: {str(e)}"
```
RequestException doesn't necessarily mean network error, it could mean invalid JSON, TooManyRedirects etc ([docs](https://requests.readthedocs.io/en/latest/_modules/requests/exceptions/)).

elazarlachkar commented on elementary/messages/messaging_integrations/teams_webhook.py:
```diff
@@ -56,22 +61,35 @@ def __init__(self, url: str) -> None:
+            # to determine if the message was sent successfully.
+            if (
+                response.status_code not in (HTTPStatus.OK, HTTPStatus.ACCEPTED)
+                or len(response.text) > 1
```
So if the response has any content, it will be treated as an error?
You wrote in the comment `in the response there is nothing that can be used`, that means there is some content which will always be treated as error, which I don't think we want.

FrankTub replied:
I tried to clarify some things in the comments and in the code. Let me know if it needs more attention. Just to give you some more context:

You have the old webhook version of Teams, in that version they always return status code of `200`. And if there is an error they add it in the response text. If there is no error the response text will be '1'.

The new webhook version of Teams is even worse, there they always return `202` and nothing in the response can be used to determine if it was actually sent.

In cases when you have no internet access or the webhook is down or something another http status code is returned.

elazarlachkar commented on elementary/monitor/data_monitoring/alerts/data_monitoring_alerts.py:
```diff
@@ -87,6 +90,11 @@ def __init__(
         self.override_config_defaults = override_config
         self.alerts_integration = self._get_integration_client()
+        self.alert_message_builder = AlertMessageBuilder(
+            MessageBuilderConfig(
+                maximum_columns_in_alert_samples=self.config.maximum_columns_in_alert_samples
+            )
+        )
```
The flag is used by our Cloud product. Your PR keeps the existing implementation in the OSS, so it's Ok :)
